libshpool/src/config_watcher.rs

```diff
@@ -0,0 +1,720 @@
+        }
+
+        #[test]
+        fn existing_file() {
+            let tmpdir = tempfile::tempdir().unwrap();
+            let target_path = tmpdir.path().join(Path::new("sub1/sub2/c.txt"));
+
+            let parent_path = target_path.parent().unwrap();
+
+            fs::create_dir_all(parent_path).unwrap();
+            fs::write(&target_path, "test").unwrap();
+
+            let mut watcher = recommended_watcher(|_| {}).unwrap();
+            let (watched_path, immediate_child) =
+                best_effort_watch(&mut watcher, &target_path).unwrap();
+
+            assert_eq!(watched_path, target_path);
+            assert_eq!(immediate_child, None);
+        }
+    }
+
+    mod handle_event {
+        use super::*;
+        use assert_matches::assert_matches;
+        use notify::{
+            event::{CreateKind, ModifyKind, RemoveKind, RenameMode},
+            Event, EventKind,
+        };
+        use ntest::test_case;
+
+        fn paths_entry(target: &str, watched: &str) -> (PathBuf, (PathBuf, PathBuf)) {
+            let target = PathBuf::from(target);
+            let base = PathBuf::from(watched);
+            let immediate =
+                base.join(target.strip_prefix(&base).unwrap().iter().next().unwrap_or_default());
+            (target, (base, immediate))
+        }
+
+        // create event from spec:
+        // <create|mv|modify|rm> path1 [path2]
+        // `base` is prepended to all paths
+        fn event_from_spec(base: &str, evt: &str) -> notify::Event {
+            let base = Path::new(base);
+            let (evt, path) = evt.split_once(' ').unwrap_or((evt, ""));
+            match evt {
+                "create" => {
+                    Event::new(EventKind::Create(CreateKind::Any)).add_path(base.join(path))
+                }
+                "mv" => {
+                    let (src, dst) = path.split_once(' ').unwrap();
+                    Event::new(EventKind::Modify(ModifyKind::Name(RenameMode::Both)))
+                        .add_path(base.join(src))
+                        .add_path(base.join(dst))
+                }
+                "mvselfother" => Event::new(EventKind::Modify(ModifyKind::Name(RenameMode::Both)))
+                    .add_path(base.to_owned())
+                    .add_path(PathBuf::from("/some/other/path")),
+                "modify" => {
+                    Event::new(EventKind::Modify(ModifyKind::Any)).add_path(base.join(path))
+                }
+                "modifyself" => {
+                    Event::new(EventKind::Modify(ModifyKind::Any)).add_path(base.to_owned())
+                }
+                "rm" => Event::new(EventKind::Remove(RemoveKind::Any)).add_path(base.join(path)),
+                "rmself" => {
+                    Event::new(EventKind::Remove(RemoveKind::Any)).add_path(base.to_owned())
+                }
+                _ => panic!("malformatted event spec"),
+            }
+        }
+
+        #[test]
```

Comment:
Just to be safe, we should probably put `#[timeout(30000)]` on all of these tests